File: src/token_goat/mcp_cache.py
"""MCP tool result cache — dedup repeated read-only MCP calls within a session.

Storage mirrors web_cache: blobs are gzip-compressed under
``data_dir() / "mcp_outputs"``.  The session carries a
``mcp_result_hashes`` dict (tool+input hash → output_id) so the
pre-fetch hook can detect repeat calls and deny them with a cached hint.
"""
from __future__ import annotations

import json
import re
import time
from pathlib import Path

from .cache_common import (
    build_output_id,
    get_cache_dir,
    load_blob_gz,
    short_content_hash,
    store_blob_gz,
)

__all__ = [
    "is_mcp_read_only",
    "mcp_hash",
    "store_mcp_result",
    "load_mcp_result",
    "MCP_MAX_CACHE_BYTES",
]

# Maximum bytes stored per MCP result blob (2 MB).
MCP_MAX_CACHE_BYTES: int = 2 * 1024 * 1024

# Blocklist of mutation verbs matched against the trailing method component of
# the tool name (e.g. "create_issue" in "mcp__plugin_github_github__create_issue").
# Uses (?:^|_)verb(?=_|$) anchoring because underscore is \w, so \b does not fire
# between a verb and the following _ separator (e.g. \bcreate\b misses create_issue).
# Assumes snake_case method names — all Claude Code / Codex CLI MCP tool registries
# use lowercase_snake_case; camelCase tools are not present in practice.
_MUTABLE_VERBS_RE = re.compile(
    r"(?:^|_)(?:create|update|delete|send|write|push|post|remove|label|unlabel|merge|"
    r"modify|draft|fork|reply|move|rename|set|add|run|execute|close|"
    r"request|upload|insert|revoke|reset|archive|restore|annotate|register|"
    r"unregister|star|unstar|like|unlike|vote|block|unblock|invite|kick|ban)(?=_|$)",
    re.IGNORECASE,
)


def is_mcp_read_only(tool_name: str) -> bool:
    """Return True when *tool_name* is a read-only MCP tool safe to cache.

    Only ``mcp__``-prefixed tools are considered.  Applies a blocklist of
    mutation verbs to the last ``__``-delimited component (the method name).
    """
    if not tool_name.startswith("mcp__"):
        return False
    method = tool_name.rsplit("__", 1)[-1]
    return not bool(_MUTABLE_VERBS_RE.search(method))


def mcp_hash(tool_name: str, tool_input: dict) -> str:  # type: ignore[type-arg]
    """Return a 16-char hex hash for the (tool_name, tool_input) pair.

    Input dict is JSON-serialized with sorted keys for stability across
    invocations that construct the same dict in different insertion orders.
    """
    canonical = json.dumps(
        {"tool": tool_name, "input": tool_input},
        sort_keys=True,
        ensure_ascii=False,
    )
    return short_content_hash(canonical)


def _mcp_outputs_dir() -> Path:
    return get_cache_dir("mcp_outputs")


def store_mcp_result(
    session_id: str,
    tool_input_hash: str,
    result_text: str,
    ts: float | None = None,
) -> str | None:
    """Write *result_text* gzip-compressed to the MCP output store.

    Returns the ``output_id`` on success, or ``None`` when the blob exceeds
    :data:`MCP_MAX_CACHE_BYTES` or the write fails.
    """
    if len(result_text.encode("utf-8", errors="replace")) > MCP_MAX_CACHE_BYTES:
        return None
    _ts = ts if ts is not None else time.time()
    output_id = build_output_id(session_id, tool_input_hash, _ts)
    path = store_blob_gz(output_id, result_text, _mcp_outputs_dir, "mcp_cache")
    return output_id if path is not None else None


def load_mcp_result(output_id: str) -> str | None:
    """Return the cached MCP result text for *output_id*, or ``None``."""
    return load_blob_gz(output_id, _mcp_outputs_dir, "mcp_cache")

File: packages/token-goat-ts/src/mcp_cache.ts
/**
 * MCP tool result cache — dedup repeated read-only MCP calls within a session.
 */

import * as fs from 'fs/promises'
import { resolve } from 'path'
import { fingerprintContent } from './fingerprint.js'
import { dataDir } from './constants.js'

export const MCP_DEFAULT_MAX_TOTAL_BYTES = 32 * 1024 * 1024
export const MCP_MAX_CACHE_BYTES = 2 * 1024 * 1024

/** Metadata associated with a cached MCP result entry. */
export interface McpOutputMeta {
  readonly outputId: string
  readonly toolName: string
  readonly inputPreview: string
  readonly resultBytes: number
  readonly ts: number
}

const MUTABLE_VERBS_RE = /(?:^|_)(?:create|update|delete|send|write|push|post|remove|label|unlabel|merge|modify|draft|fork|reply|move|rename|set|add|run|execute|close|copy|request|upload|insert|revoke|reset|archive|restore|annotate|register|unregister|star|unstar|like|unlike|vote|block|unblock|invite|kick|ban)(?=_|$)/i

const _resultsByHash = new Map<string, { text: string; ts: number }>()
const _metaByOutputId = new Map<string, McpOutputMeta>()

/**
 * Return True when *toolName* is a read-only MCP tool safe to cache.
 * Only `mcp__`-prefixed tools are considered.
 */
export function isMcpReadOnly(toolName: string): boolean {
  if (!toolName.startsWith('mcp__')) {
    return false
  }
  const method = toolName.split('__').pop() || ''
  return !MUTABLE_VERBS_RE.test(method)
}

/**
 * Return a 16-char hex hash for the (toolName, toolInput) pair.
 * Input dict is JSON-serialized with sorted keys for stability.
 */
export function mcpHash(toolName: string, toolInput: Record<string, unknown>): string {
  const sortedInput: Record<string, unknown> = {}
  for (const key of Object.keys(toolInput).sort()) {
    sortedInput[key] = toolInput[key]
  }
  const canonical = JSON.stringify({ tool: toolName, input: sortedInput })
  return fingerprintContent(canonical).slice(0, 16)
}

/**
 * Return the sidecar JSON metadata path for *outputId*.
 */
export function sidecarMetaPath(outputId: string): string | null {
  if (!outputId || outputId.includes('..') || outputId.includes('/')) {
    return null
  }
  const baseDir = resolve(dataDir(), 'mcp_outputs')
  return resolve(baseDir, `${outputId}.json`)
}

/**
 * Write *meta* as a JSON sidecar (best-effort).
 */
export async function writeSidecar(meta: McpOutputMeta): Promise<void> {
  try {
    const path = sidecarMetaPath(meta.outputId)
    if (!path) return
    const dir = resolve(path, '..')
    await fs.mkdir(dir, { recursive: true })
    await fs.writeFile(path, JSON.stringify(meta, null, 2) + '\n')
    _metaByOutputId.set(meta.outputId, meta)
  } catch {
    // Best-effort; swallow errors
  }
}

/**
 * Return parsed McpOutputMeta from cache or sidecar, or null.
 */
export async function readSidecar(outputId: string): Promise<McpOutputMeta | null> {
  if (_metaByOutputId.has(outputId)) {
    return _metaByOutputId.get(outputId) || null
  }

  try {
    const path = sidecarMetaPath(outputId)
    if (!path) return null
    const content = await fs.readFile(path, 'utf-8')
    const data = JSON.parse(content)
    const meta: McpOutputMeta = {
      outputId: String(data.outputId || outputId),
      toolName: String(data.toolName || ''),
      inputPreview: String(data.inputPreview || ''),
      resultBytes: Number(data.resultBytes || 0),
      ts: Number(data.ts || 0),
    }
    _metaByOutputId.set(outputId, meta)
    return meta
  } catch {
    return null
  }
}

/**
 * Write *resultText* to the MCP output store and return the outputId, or null on error.
 * Returns null when the blob exceeds MCP_MAX_CACHE_BYTES or the write fails.
 */
export async function storeMcpResult(
  sessionId: string,
  toolInputHash: string,
  resultText: string,
  ts?: number,
  opts?: { toolName?: string; inputPreview?: string }
): Promise<string | null> {
  const resultBytes = Buffer.byteLength(resultText, 'utf-8')
  if (resultBytes > MCP_MAX_CACHE_BYTES) {
    return null
  }

  const timestamp = ts ?? Date.now()
  const outputId = `${sessionId}_${toolInputHash}_${Math.floor(timestamp / 1000)}`

  try {
    const dir = resolve(dataDir(), 'mcp_outputs')
    await fs.mkdir(dir, { recursive: true })
    _resultsByHash.set(toolInputHash, { text: resultText, ts: timestamp })

    if (opts?.toolName) {
      const meta: McpOutputMeta = {
        outputId,
        toolName: opts.toolName,
        inputPreview: (opts.inputPreview || '').slice(0, 200),
        resultBytes,
        ts: timestamp,
      }
      await writeSidecar(meta)
    }

    return outputId
  } catch {
    return null
  }
}

/**
 * Return the cached MCP result text for *outputId*, or null.
 */
export async function getMcpResult(sessionId: string, toolInputHash: string): Promise<string | null> {
  if (_resultsByHash.has(toolInputHash)) {
    return _resultsByHash.get(toolInputHash)?.text || null
  }
  return null
}

/**
 * Clear all in-memory caches (for testing).
 */
export function reset(): void {
  _resultsByHash.clear()
  _metaByOutputId.clear()
}
